src/main/java/com/artozersky/HackerNewsAPI/service/impl/NewsPostServiceImpl.java

```diff
@@ -0,0 +1,262 @@
+    }
+    
+    @Override
+    public  NewsPostsResponseDTOImpl getPostById(Long postId) {
+
+        NewsPostModelImpl cachedValue = cacheService.getPostFromCacheById(postId);
+
+        if (cachedValue != null) {
+            return convertToDTO(cachedValue);
+        }
+
+        NewsPostModelImpl postModel = postRepository.findById(postId)
+            .orElseThrow(() -> new CustomNotFoundException("Post not found with id: " + postId));
+
+        //cacheService.putPostInCache(postModel);
+        //should i update the cache with the top posts???
+
+        return convertToDTO(postModel);
+    }
+
+    @Override
+    public List<NewsPostsResponseDTOImpl> getAllPosts(int limit) {
+
+        if (limit <= 0) {
+            throw new CustomServiceException("Limit must be greater than 0");
+        }
+
+        Pageable pageable = PageRequest.of(0, limit);
+        Page<NewsPostModelImpl> page = postRepository.findAll(pageable);
+
+        return page.getContent().stream()
+                .map(this::convertToDTO)
+                .collect(Collectors.toList());
+    }
+
+    @Override
+    public List<NewsPostsResponseDTOImpl> getTopPosts(int limit) {
+
+    List<NewsPostModelImpl> cachedPosts = cacheService.getAllPostsFromCache();//stale will be checked inside
+
+    int postsNeeded = limit - cachedPosts.size();
+
+    System.out.println("cached posts size: " + cachedPosts.size() );
+
+    System.out.println("posts needed: " + postsNeeded );
+
+    List<NewsPostModelImpl> dbPosts = new ArrayList<>();
+
+    dbPosts = fetchPostsFromDbIfNeeded(postsNeeded, cachedPosts);
+
+    System.out.println("fetched from db: " + dbPosts.size() );
+
+    List<NewsPostModelImpl> combinedPosts = new ArrayList<>(cachedPosts);
+    combinedPosts.addAll(dbPosts);
+
+    return combinedPosts.stream()
+        .map(this::convertToDTO)
+        .collect(Collectors.toList());
+}
+
+    
+    @Override
+    public NewsPostsResponseDTOImpl savePost(@Valid NewsPostsCreateDTOImpl postCreateDTO) {
+        
+        NewsPostModelImpl post = modelMapper.map(postCreateDTO, NewsPostModelImpl.class);
+        
+        post.initialize();
+        
+        NewsPostModelImpl savedPost = postRepository.save(post);
+        
+        NewsPostsResponseDTOImpl responseDTO = modelMapper.map(savedPost, NewsPostsResponseDTOImpl.class);
+        
+        responseDTO.setMessage("Post created successfully");
+
+        updateCacheWithTopPosts();
+        
+        return responseDTO;
+    }
+
+    @Override
+    public NewsPostsResponseDTOImpl updatePost(NewsPostsUpdateDTOImpl postUpdateDTO, Long postId) {
+        
+        NewsPostModelImpl post = postRepository.findById(postId)
+        .orElseThrow(() -> new CustomNotFoundException("Post not found with id: " + postId));
+        
+        boolean isRedundant = true;
+        
+        if (postUpdateDTO.getTitle() != null && !postUpdateDTO.getTitle().trim().isEmpty()) {
```

Comment:
all those if statements can be saved because you can enforce in the model via the annotation that a url/title etc can not be null and empty. this is a must and then if the user will try empty or null it will throw an exception which will get to the global exception manager and then it will raise error accordingly.